temporalio/worker/workflow.py

```diff
@@ -300,3 +338,22 @@ async def _create_workflow_instance(
                 extern_functions=self._extern_functions,
             )
         )
+
+
+class _UnexpectedEvictionError(temporalio.exceptions.TemporalError):
+    def __init__(
+        self,
+        reason: temporalio.bridge.proto.workflow_activation.RemoveFromCache.EvictionReason.ValueType,
+        message: str,
+    ) -> None:
+        self.reason = temporalio.bridge.proto.workflow_activation.RemoveFromCache.EvictionReason.Name(
+            reason
+        )
+        self.message = message
+        super().__init__(f"{self.reason}: {message}")
+
+
+# TODO(cretz): Make this public and move to a more centralized place whenever a
+# sandbox is around that could justify its reuse
+class _NondeterminismError(_UnexpectedEvictionError):
+    pass
```

Comment:
Seems to me it makes sense to already make this public. Even outside of the replay context this error can already get thrown, so might as well bubble it up just like is happening in your tests here.

Reply:
I couldn't decide where to put it. TypeScript put it at https://typescript.temporal.io/api/classes/workflow.DeterminismViolationError, so I can technically put it in https://python.temporal.io/temporalio.workflow.html, but I also would want it to inherit from `temporalio.exceptions.TemporalError` which I don't do for anything else in that `temporalio.workflow` package today (that package is meant to only be called from workflows).

I just decided to punt, but will take a vote:

1. `temporalio.workflow.NondeterminismError`
2. `temporalio.workflow.DeterminismViolationError`
3. `temporalio.exceptions.NondeterminismError`
4. `temporalio.exceptions.DeterminismViolationError`

Reply:
I vote 1

Reply:
I'll leave this up for a day or two then switch to 1. The `temporalio.workflow` package was meant only for user workflow code (whether defining or running), but I can change it a bit for this one use case.

Reply:
Switched to 1